### PR TITLE
[neox] persistence

### PR DESCRIPTION
**Pull request by KickSeason**
### mpt integration (Step 2/4) 
* Add `LocalStateRoot` in persistence to store every local root hash when block persisted completed.
* Add `ValidatorsStateRoot ` to store the latest confirmed state root with cn signatures.
* Use MPT in `Storages`
  
We use the latest `StateRoot` from cn to verify local state no matter historical state roots are.

### CODE REVIEW

File: src/neo/Persistence/StoreView.cs [hunk 1]
```diff
@@ -14,17 +15,21 @@ public abstract class StoreView
...
         public abstract DataCache<UInt256, TransactionState> Transactions { get; }
         public abstract DataCache<UInt160, ContractState> Contracts { get; }
-        public abstract DataCache<StorageKey, StorageItem> Storages { get; }
         public abstract DataCache<SerializableWrapper<uint>, HeaderHashList> HeaderHashList { get; }
+        public abstract DataCache<SerializableWrapper<uint>, HashState> LocalStateRoot { get; }
         public abstract MetaDataCache<HashIndexState> BlockHashIndex { get; }
         public abstract MetaDataCache<HashIndexState> HeaderHashIndex { get; }
+        public abstract MetaDataCache<StateRoot> ConfirmedStateRootHashIndex { get; }
         public abstract MetaDataCache<ContractIdState> ContractId { get; }
+        public MPTTrie<StorageKey, StorageItem> Storages;
```
Comment (shargon) [inline src/neo/Persistence/StoreView.cs:24]: Why not abstract?

File: tests/neo.UnitTests/Network/P2P/Payloads/UT_Transaction.cs [hunk 2]
```diff
@@ -692,11 +691,10 @@ public void FeeIsSignatureContract_TestScope_NoScopeFAULT()
                 // Fake balance
 
                 var key = NativeContract.GAS.CreateStorageKey(20, acc.ScriptHash);
-
-                var entry = snapshot.Storages.GetAndChange(key, () => new StorageItem(new AccountState()));
```
Comment (shargon) [inline tests/neo.UnitTests/Network/P2P/Payloads/UT_Transaction.cs:696]: Why remove this UT?

Comment (KickSeason) [reply tests/neo.UnitTests/Network/P2P/Payloads/UT_Transaction.cs:696]: `MPTTrie` doesn't have a change set.

**Comment by roman-khimov** [comment]
I don't think this one takes into account the discussion in #1526 and #1383 and follows the design we currently have in neox-2.x with completely detached StateRoot. While this design might be appropriate for 2.x because of backwards compatibility requirements, it's certainly not optimal for 3.0 because as we have recently seen in the testnet, having block sync detached from state sync is dangerous and better be avoided.

Putting StateRoot in the block header also simplifies the design a lot, we just won't need any additional messages and won't have the situation when there is a state, but it's 'unconfirmed'.

You can also look at it from a bit different perspective --- this network is synchronized by blocks and it shouldn't require anything additional for correct operation, while a separate StateRoot entities distributed with different messages create an additional synchronization mechanism that at the same time somehow affects the primary block-bound synchronization. This leads to complex interaction between these two mechanism and creates additional failure modes that again can be avoided with a simple move of StateRoot info into the block header.

**Comment by KickSeason** [comment]
@roman-khimov We should also consider upgrade and rollback.
Agree what we have learnt in testnet. I think we can continue improving it. Anyway it's only our preview1.
On neo3, we can use the latest state root instead sync one by one. Remove the state height requirement for consensus node. So that we can avoid the influence to block consensus.
I agree put state root into header is simple and clean. But we need  virtual execution environments to run old blocks. We need a specific solution to this before we put state root into header.

**Comment by roman-khimov** [comment]
> We should also consider upgrade and rollback.

We have #1287 and neo-project/neo-vm#142 proposals that can handle that in a well-defined predictable manner.

> Agree what we have learnt in testnet. I think we can continue improving it. Anyway it's only our preview1.

Well, that's kinda what testnet and previews are made for --- to see if and how things break with more realistic usage scenarios. But at the same time I think we should immediately take into account the experience we're getting from these tests and previews. And that experience shows some problems with the design we currently have.

> We need a specific solution to this before we put state root into header.

That's probably where I would disagree, we can make an implementation of header-tied StateRoot first and deal with versioning/upgrade handling later, there is nothing preventing that. At the same time it could also be done the other way around starting with versioning, of course. But any of these two routes saves us some time on implementing/debugging/reviewing/reworking StateRoot messages and block-state synchronization issues.

**Comment by KickSeason** [comment]
@erikzhang  Can you have a look?

File: src/neo/Persistence/ReadOnlyView.cs [hunk 3]
```diff
@@ -15,11 +17,23 @@ public class ReadOnlyView : StoreView
...
         public override MetaDataCache<ContractIdState> ContractId => new StoreMetaDataCache<ContractIdState>(store, Prefixes.IX_ContractId);
+        public override MPTTrie<StorageKey, StorageItem> Storages
+        {
+            get
+            {
+                return new MPTTrie<StorageKey, StorageItem>((ISnapshot)store, CurrentStateRootHash);
+            }
+            set
+            {
+                Storages = value;
```
Comment (shargon) [inline src/neo/Persistence/ReadOnlyView.cs:34]: infinity loop?

File: src/neo/Network/P2P/Payloads/StateRoot.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            }
+            set
+            {
+                if (value.Length != 1) throw new ArgumentException();
+                Witness = value[0];
+            }
+        }
+
+        public int Size =>
+            sizeof(byte) +
```
Comment (shargon) [inline src/neo/Network/P2P/Payloads/StateRoot.cs:50]: We usually add comments with the name of the variable

File: src/neo/Ledger/HashState.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
+using Neo.IO;
+using Neo.Network.P2P.Payloads;
+using System.IO;
+
+namespace Neo.Ledger
+{
+    public class HashState : ICloneable<HashState>, ISerializable
```
Comment (Tommo-L) [inline src/neo/Ledger/HashState.cs:7]: Need a better name.

File: src/neo/Persistence/ClonedView.cs [hunk 6]
```diff
@@ -1,31 +1,45 @@
...
 using Neo.IO;
 using Neo.IO.Caching;
 using Neo.Ledger;
+using Neo.Network.P2P.Payloads;
 
 namespace Neo.Persistence
 {
     internal class ClonedView : StoreView
     {
+        private StoreView parent;
```
Comment (Tommo-L) [inline src/neo/Persistence/ClonedView.cs:11]: It may not a good idea by using parent storeview.

File: src/neo/Persistence/SnapshotView.cs [hunk 7]
```diff
@@ -15,27 +17,41 @@ public class SnapshotView : StoreView, IDisposable
...
+            ConfirmedStateRoot = new StoreMetaDataCache<StateRoot>(snapshot, Prefixes.IX_ConfirmedRoot);
+            Storages = new MPTTrie<StorageKey, StorageItem>(snapshot, CurrentStateRootHash);
+        }
+
+        private void PreCommit()
+        {
+            var index = PersistingBlock?.Index ?? Height;
+            if (LocalStateRoot.TryGet(index) is null || LocalStateRoot.TryGet(index).Hash != Storages.Root.Hash)
+            {
+                LocalStateRoot.GetAndChange(index, () => new HashState()).Hash = Storages.Root.Hash;
```
Comment (Tommo-L) [inline src/neo/Persistence/SnapshotView.cs:48]: Maybe need a better place to do it.

**Comment by Tommo-L** [comment]
@KickSeason plz, help fix the ut.

File: src/neo/Network/P2P/Payloads/StateRoot.cs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        public void SerializeUnsigned(BinaryWriter writer)
+        {
+            writer.Write(Version);
+            writer.Write(Index);
+            writer.Write(Root);
+        }
+
+        public bool Verify(StoreView snapshot)
+        {
+            return this.VerifyWitnesses(snapshot,
```
Comment (Tommo-L) [inline src/neo/Network/P2P/Payloads/StateRoot.cs:102]:
```suggestion
            return this.VerifyWitnesses(snapshot, 1_00000000)
```

Keep the same with BlockBase.Verify

https://github.com/neo-project/neo/blob/981ba21c0d2b7e67dab539213f71d36dcc33e775/src/neo/Network/P2P/Payloads/BlockBase.cs#L121

File: src/neo/Persistence/Prefixes.cs [hunk 9]
```diff
@@ -7,11 +7,14 @@ internal static class Prefixes
 
         public const byte ST_Contract = 0x50;
         public const byte ST_Storage = 0x70;
+        public const byte ST_Root = 0x60;
```
Comment (Tommo-L) [inline src/neo/Persistence/Prefixes.cs:10]:
```suggestion
        public const byte ST_LocalRoot = 0x60;
```

File: src/neo/Persistence/Prefixes.cs [hunk 10]
```diff
@@ -7,11 +7,14 @@ internal static class Prefixes
...
 
         public const byte IX_HeaderHashList = 0x80;
         public const byte IX_CurrentBlock = 0xc0;
         public const byte IX_CurrentHeader = 0xc1;
         public const byte IX_ContractId = 0xc2;
+        public const byte IX_CurrentRoot = 0xc3;
+        public const byte IX_ConfirmedRoot = 0xc4;
```
Comment (Tommo-L) [inline src/neo/Persistence/Prefixes.cs:17]:
```suggestion
        public const byte IX_ValidatorRoot = 0xc4;
```

Comment (Tommo-L) [reply src/neo/Persistence/Prefixes.cs:17]: Delete `IX_CurrentRoot`.

File: src/neo/Persistence/ReadOnlyView.cs [hunk 11]
```diff
@@ -15,11 +17,13 @@ public class ReadOnlyView : StoreView
...
         public override DataCache<SerializableWrapper<uint>, HeaderHashList> HeaderHashList => new StoreDataCache<SerializableWrapper<uint>, HeaderHashList>(store, Prefixes.IX_HeaderHashList);
+        public override DataCache<SerializableWrapper<uint>, HashIndexState> LocalStateRoot => new StoreDataCache<SerializableWrapper<uint>, HashIndexState>(store, Prefixes.ST_Root);
         public override MetaDataCache<HashIndexState> BlockHashIndex => new StoreMetaDataCache<HashIndexState>(store, Prefixes.IX_CurrentBlock);
         public override MetaDataCache<HashIndexState> HeaderHashIndex => new StoreMetaDataCache<HashIndexState>(store, Prefixes.IX_CurrentHeader);
+        public override MetaDataCache<StateRoot> LatestVerifiedStateRoot => new StoreMetaDataCache<StateRoot>(store, Prefixes.IX_ConfirmedRoot);
         public override MetaDataCache<ContractIdState> ContractId => new StoreMetaDataCache<ContractIdState>(store, Prefixes.IX_ContractId);
+        public override DataCache<StorageKey, StorageItem> Storages => new MPTDataCache<StorageKey, StorageItem>(store, Prefixes.ST_Storage, CurrentStateRootHash);
```
Comment (Tommo-L) [inline src/neo/Persistence/ReadOnlyView.cs:26]: Add comments for `Storages`.

File: src/neo/Cryptography/MPT/MPTTrie.Proof.cs [hunk 12]
```diff
@@ -63,9 +63,9 @@ public static TValue VerifyProof(UInt256 root, TKey key, HashSet<byte[]> proof)
         {
             using var memoryStore = new MemoryStore();
             foreach (byte[] data in proof)
-                memoryStore.Put(Prefix, Crypto.Hash256(data), data);
+                memoryStore.Put(0, Crypto.Hash256(data), data);
             using ISnapshot snapshot = memoryStore.GetSnapshot();
-            var trie = new MPTTrie<TKey, TValue>(snapshot, root);
+            var trie = new MPTTrie<TKey, TValue>(snapshot, 0, root);
```
Comment (Tommo-L) [inline src/neo/Cryptography/MPT/MPTTrie.Proof.cs:68]: Replace `0` by a const variable

File: src/neo/Persistence/StoreView.cs [hunk 13]
```diff
@@ -14,15 +14,18 @@ public abstract class StoreView
         public abstract DataCache<UInt256, TrimmedBlock> Blocks { get; }
         public abstract DataCache<UInt256, TransactionState> Transactions { get; }
         public abstract DataCache<UInt160, ContractState> Contracts { get; }
-        public abstract DataCache<StorageKey, StorageItem> Storages { get; }
         public abstract DataCache<SerializableWrapper<uint>, HeaderHashList> HeaderHashList { get; }
+        public abstract DataCache<SerializableWrapper<uint>, HashIndexState> LocalStateRoot { get; }
         public abstract MetaDataCache<HashIndexState> BlockHashIndex { get; }
         public abstract MetaDataCache<HashIndexState> HeaderHashIndex { get; }
+        public abstract MetaDataCache<StateRoot> LatestVerifiedStateRoot { get; }
```
Comment (Tommo-L) [inline src/neo/Persistence/StoreView.cs:21]:
```suggestion
        public abstract MetaDataCache<StateRoot> ValidatorStateRoot { get; }
```

File: src/neo/Persistence/SnapshotView.cs [hunk 14]
```diff
@@ -15,28 +16,36 @@ public class SnapshotView : StoreView, IDisposable
...
+            LatestVerifiedStateRoot = new StoreMetaDataCache<StateRoot>(snapshot, Prefixes.IX_ConfirmedRoot);
+            Storages = new MPTDataCache<StorageKey, StorageItem>(snapshot, Prefixes.ST_Storage, CurrentStateRootHash);
         }
 
         public override void Commit()
         {
             base.Commit();
+            var root = LocalStateRoot.GetAndChange(Height, () => new HashIndexState());
+            root.Index = Height;
+            root.Hash = ((MPTDataCache<StorageKey, StorageItem>)Storages).Root.Hash;
```
Comment (Tommo-L) [inline src/neo/Persistence/SnapshotView.cs:47]: Move it to blockchain.cs

**Comment by erikzhang** [comment]
Update the consensus first?

**Comment by KickSeason** [comment]
> Update the consensus first?

Will  commit  consensus change in another soon.

File: src/neo/Cryptography/MPT/MPTTrie.Proof.cs [hunk 15]
```diff
@@ -61,11 +61,12 @@ private bool GetProof(ref MPTNode node, ReadOnlySpan<byte> path, HashSet<byte[]>
 
         public static TValue VerifyProof(UInt256 root, TKey key, HashSet<byte[]> proof)
         {
+            const byte proof_prefix = 0x00;
```
Comment (shargon) [inline src/neo/Cryptography/MPT/MPTTrie.Proof.cs:64]: outside the method?

File: src/neo/Cryptography/MPT/MPTTrie.cs [hunk 16]
```diff
@@ -8,17 +8,18 @@ public partial class MPTTrie<TKey, TValue>
         where TKey : notnull, ISerializable, new()
         where TValue : class, ISerializable, new()
     {
-        private const byte Prefix = 0xf0;
+        private byte Prefix;
```
Comment (erikzhang) [inline src/neo/Cryptography/MPT/MPTTrie.cs:11]: Why this is changed?

Comment (Tommo-L) [reply src/neo/Cryptography/MPT/MPTTrie.cs:11]: If we have multiple MPTs, then `Prefix` need to be passed by the outside. And we have defined all the prefixes of table in `Prefixes.cs` file.

Comment (KickSeason) [reply src/neo/Cryptography/MPT/MPTTrie.cs:11]: What is the benefit using hard code.

Comment (erikzhang) [reply src/neo/Cryptography/MPT/MPTTrie.cs:11]: > If we have multiple MPTs

Why do we need multiple MPTs?

Comment (Tommo-L) [reply src/neo/Cryptography/MPT/MPTTrie.cs:11]: As neo-core is the library, some other clients may want to use the MPT.

Comment (erikzhang) [reply src/neo/Cryptography/MPT/MPTTrie.cs:11]: Why does a user care about what prefix should be used? I think it is a implementation detail. A caller care about `key`, `value`, but not `prefix`.

Comment (Tommo-L) [reply src/neo/Cryptography/MPT/MPTTrie.cs:11]: okay.

File: src/neo/Persistence/Helper.cs [hunk 17]
```diff
@@ -8,5 +9,13 @@ public static byte[] EnsureNotNull(this byte[] source)
         {
             return source ?? Array.Empty<byte>();
         }
+
+        public static void UpdateLocalStateRoot(this SnapshotView snapshot)
+        {
+            snapshot.Storages.Commit();
+            var root = snapshot.LocalStateRoot.GetAndChange(snapshot.Height, () => new HashIndexState());
+            root.Index = snapshot.Height;
+            root.Hash = ((MPTDataCache<StorageKey, StorageItem>)snapshot.Storages).Root.Hash;
```
Comment (shargon) [inline src/neo/Persistence/Helper.cs:18]: If we require a `MPTDataCache` maybe we should modify https://github.com/neo-project/neo/pull/1692/files#diff-de4093105173aee33135b364a5a08edaR17 to be a `MPTDataCache`

Comment (Tommo-L) [reply src/neo/Persistence/Helper.cs:18]: But we can't convert `DataClone` to `MPTCache` after `CreateSnapshot` method.
https://github.com/neo-project/neo/blob/835fdeeca87d909fce9b405160102d3250d1bed0/src/neo/IO/Caching/DataCache.cs#L104-L107

**Comment by shargon** [comment]
Could you merge this neox changes in other branch different than `master` ? we need the specification and the agreement of https://github.com/neo-project/neo/issues/1526 first , so it will be faster if we join the changes, but not merge it into `master` until these decisions.

**Comment by KickSeason** [comment]
@shargon  Can you create `neox` branch?

**Comment by shargon** [comment]
@KickSeason done https://github.com/neo-project/neo/tree/neox

**Comment by KickSeason** [comment]
Close this to change to `neox`.

**Pull request closed without merging**